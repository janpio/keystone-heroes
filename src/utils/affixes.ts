export const affixes = [
  {
    name: "Volcanic",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/spell_shaman_lavasurge.jpg",
    id: 3,
  },
  {
    name: "Necrotic",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/spell_deathknight_necroticplague.jpg",
    id: 4,
  },
  {
    name: "Raging",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/ability_warrior_focusedrage.jpg",
    id: 6,
  },
  {
    name: "Bolstering",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/ability_warrior_battleshout.jpg",
    id: 7,
  },
  {
    name: "Sanguine",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/spell_shadow_bloodboil.jpg",
    id: 8,
  },
  {
    name: "Tyrannical",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/achievement_boss_archaedas.jpg",
    id: 9,
  },
  {
    name: "Fortified",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/ability_toughness.jpg",
    id: 10,
  },
  {
    name: "Bursting",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/ability_ironmaidens_whirlofblood.jpg",
    id: 11,
  },
  {
    name: "Grievous",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/ability_backstab.jpg",
    id: 12,
  },
  {
    name: "Explosive",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/spell_fire_felflamering_red.jpg",
    id: 13,
  },
  {
    name: "Quaking",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/spell_nature_earthquake.jpg",
    id: 14,
  },
  {
    name: "Prideful",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/spell_animarevendreth_buff.jpg",
    id: 121,
  },
  {
    name: "Inspiring",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/spell_holy_prayerofspirit.jpg",
    id: 122,
  },
  {
    name: "Spiteful",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/spell_holy_prayerofshadowprotection.jpg",
    id: 123,
  },
  {
    name: "Storming",
    icon:
      "https://assets.rpglogs.com/img/warcraft/abilities/spell_nature_cyclone.jpg",
    id: 124,
  },
] as const;
