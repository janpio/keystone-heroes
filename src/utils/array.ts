export const toUniqueArray = <Dataset>(array: Dataset[]): Dataset[] => [
  ...new Set(array),
];
