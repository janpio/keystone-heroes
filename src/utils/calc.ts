import type {
  HealingDone,
  DamageDone,
  DamageTaken,
  InDepthCharacterInformation,
} from "../server/queries/table";
import type { Dungeon } from "./dungeons";

export const calcMetricAverage = <
  T extends Pick<DamageDone | HealingDone | DamageTaken, "total" | "guid">[]
>(
  keystoneTime: number,
  data: T,
  guid?: InDepthCharacterInformation["guid"]
): number => {
  const runInSeconds = keystoneTime / 1000;

  const sum = guid
    ? data.find((event) => event.guid === guid)?.total ?? 0
    : data.reduce((acc: number, dataset) => acc + dataset.total, 0);

  return Math.round(sum / runInSeconds);
};

export const calcTimeLeft = (
  { timer }: Dungeon,
  keystoneTime: number
): string => {
  const [maxTime] = timer;

  return calcRunDuration(maxTime - keystoneTime);
};

export const calcRunDuration = (keystoneTime: number): string => {
  const totalSeconds = keystoneTime / 1000;

  const minutes = Math.floor(totalSeconds / 60);
  const seconds = Math.floor(totalSeconds - minutes * 60);
  const milliseconds = keystoneTime - minutes * 60 * 1000 - seconds * 1000;

  const paddedMinutes = minutes.toString().padStart(2, "0");
  const paddedSeconds = seconds.toString().padStart(2, "0");
  const paddedMs = milliseconds.toString().padStart(3, "0");

  return `${paddedMinutes}:${paddedSeconds}.${paddedMs}`;
};
