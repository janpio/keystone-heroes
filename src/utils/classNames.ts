export const classnames = (...classes: unknown[]): string =>
  classes.filter(Boolean).join(" ");
