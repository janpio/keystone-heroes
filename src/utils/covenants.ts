export const covenantMap = {
  1: {
    icon: "//assets.rpglogs.com/img/warcraft/abilities/ui_sigil_kyrian.jpg",
    name: "Kyrian",
  },
  2: {
    icon: "//assets.rpglogs.com/img/warcraft/abilities/ui_sigil_venthyr.jpg",
    name: "Venthyr",
  },
  3: {
    icon: "//assets.rpglogs.com/img/warcraft/abilities/ui_sigil_nightfae.jpg",
    name: "Night Fae",
  },
  4: {
    icon: "//assets.rpglogs.com/img/warcraft/abilities/ui_sigil_necrolord.jpg",
    name: "Necrolord",
  },
} as const;

export type Covenants = typeof covenantMap;
export type Covenant = Covenants[keyof Covenants];

export const soulbindMap = {
  1: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-1.jpg",
    name: "",
  },
  2: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-2.jpg",
    name: "",
  },
  3: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-3.jpg",
    name: "",
  },
  4: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-4.jpg",
    name: "",
  },
  5: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-5.jpg",
    name: "",
  },
  6: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-6.jpg",
    name: "",
  },
  7: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-7.jpg",
    name: "Pelagos",
  },
  8: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-8.jpg",
    name: "",
  },
  9: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-9.jpg",
    name: "",
  },
  10: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-10.jpg",
    name: "",
  },
  11: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-11.jpg",
    name: "",
  },
  13: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-13.jpg",
    name: "Kleia",
  },
  18: {
    icon: "//assets.rpglogs.com/img/warcraft/soulbinds/soulbind-18.jpg",
    name: "",
  },
};

export type Soulbinds = typeof soulbindMap;
export type Soulbind = Soulbinds[keyof Soulbinds];
