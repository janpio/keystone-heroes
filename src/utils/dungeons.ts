import type { InitialFightInformation } from "../types";

const createTimer = (initialTime: number): [number, number, number] => [
  initialTime * 60 * 1000,
  initialTime * 60 * 1000 * 0.8,
  initialTime * 60 * 1000 * 0.6,
];

export const dungeons = [
  { id: 12_284, name: "Sanguine Depths", timer: createTimer(41), slug: "SD" },
  {
    id: 12_285,
    name: "Spires of Ascension",
    timer: createTimer(39),
    slug: "SoA",
  },
  { id: 12_286, name: "The Necrotic Wake", timer: createTimer(36), slug: "NW" },
  {
    id: 12_287,
    name: "Halls of Atonement",
    timer: createTimer(31),
    slug: "HoA",
  },
  { id: 12_289, name: "Plaguefall", timer: createTimer(38), slug: "PF" },
  {
    id: 12_290,
    name: "Mists of Tirna Scithe",
    timer: createTimer(30),
    slug: "MoTS",
  },
  { id: 12_291, name: "De Other Side", timer: createTimer(43), slug: "DOS" },
  { id: 12_293, name: "Theatre of Pain", timer: createTimer(37), slug: "TOP" },
] as const;

/**
 * `completionTime` is only larger than 0 if the key was indeed finished
 * otherwise, use log `start`/`end` timers
 */
export const getRunTime = ({
  completionTime,
  start_time,
  end_time,
}: InitialFightInformation): string => {
  const totalMilliseconds =
    completionTime > 0 ? completionTime : end_time - start_time;
  const totalSeconds = totalMilliseconds / 1000;

  const minutes = Math.floor(totalSeconds / 60);
  const seconds = Math.floor(totalSeconds - minutes * 60);
  const milliseconds = totalMilliseconds - minutes * 60 * 1000 - seconds * 1000;

  const paddedMinutes = minutes.toString().padStart(2, "0");
  const paddedSeconds = seconds.toString().padStart(2, "0");
  const paddedMs = milliseconds.toString().padStart(3, "0");

  return `${paddedMinutes}:${paddedSeconds}.${paddedMs}`;
};
