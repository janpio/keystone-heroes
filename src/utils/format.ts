import type { RawFight } from "../server/queries/report";

export const formatTimeLeft = (
  timer: [number, number, number],
  completionTime: number
): string => {
  const [maxTime] = timer;

  return formatKeystoneTime(maxTime - completionTime);
};

export const formatKeystoneTime = (
  completionTime: RawFight["keystoneTime"]
): string => {
  const totalSeconds = completionTime / 1000;

  const minutes = Math.floor(totalSeconds / 60);
  const seconds = Math.floor(totalSeconds - minutes * 60);
  const milliseconds = completionTime - minutes * 60 * 1000 - seconds * 1000;

  const paddedMinutes = minutes.toString().padStart(2, "0");
  const paddedSeconds = seconds.toString().padStart(2, "0");
  const paddedMs = milliseconds.toString().padStart(3, "0");

  return `${paddedMinutes}:${paddedSeconds}.${paddedMs}`;
};
