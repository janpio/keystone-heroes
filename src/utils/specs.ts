export const tankSpecs = new Set([
  "DemonHunter-Vengeance",
  "Warrior-Protection",
  "Paladin-Protection",
  "Monk-Brewmaster",
  "Druid-Guardian",
  "DeathKnight-Blood",
]);

export const healSpecs = new Set([
  "Druid-Restoration",
  "Shaman-Restoration",
  "Monk-Mistweaver",
  "Paladin-Holy",
  "Priest-Holy",
  "Priest-Discipline",
]);
