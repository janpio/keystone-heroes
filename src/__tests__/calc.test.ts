import type { DamageDone } from "../server/queries/fights";
import { calcMetricAverage } from "../utils/calc";

describe("calcMetricAverage", () => {
  test.each([
    0,
    10,
    100,
    1000,
    10_000,
    100_000,
    1_000_000,
    200 * 1000,
    25 * 1000 * 20,
    28 * 60 * 1000,
  ])("matches snapshot", (runDuration) => {
    const total = runDuration > 1 * 60 * 1000 ? 200 * 1000 : 200;

    const damageDone: DamageDone[] = [
      { guid: 1, icon: "", id: 1, name: "", total, type: "DemonHunter" },
      { guid: 1, icon: "", id: 1, name: "", total, type: "DemonHunter" },
      { guid: 1, icon: "", id: 1, name: "", total, type: "DemonHunter" },
      { guid: 1, icon: "", id: 1, name: "", total, type: "DemonHunter" },
      { guid: 1, icon: "", id: 1, name: "", total, type: "DemonHunter" },
    ];

    expect(calcMetricAverage(runDuration, damageDone)).toMatchSnapshot();
  });
});
