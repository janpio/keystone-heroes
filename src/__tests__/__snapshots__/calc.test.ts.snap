// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calcGroupMetricAverage matches snapshot 1`] = `Infinity`;

exports[`calcGroupMetricAverage matches snapshot 2`] = `100000`;

exports[`calcGroupMetricAverage matches snapshot 3`] = `10000`;

exports[`calcGroupMetricAverage matches snapshot 4`] = `1000`;

exports[`calcGroupMetricAverage matches snapshot 5`] = `100`;

exports[`calcGroupMetricAverage matches snapshot 6`] = `10000`;

exports[`calcGroupMetricAverage matches snapshot 7`] = `1000`;

exports[`calcGroupMetricAverage matches snapshot 8`] = `5000`;

exports[`calcGroupMetricAverage matches snapshot 9`] = `2000`;

exports[`calcGroupMetricAverage matches snapshot 10`] = `595`;
